src/diagnostics.rs

```diff
@@ -57,35 +57,69 @@ pub fn publish_diagnostics(params: Params, ctx: &mut Context) {
             )
         })
         .join(" ");
-    let mut lines_with_errors = HashSet::new();
-    let diagnostic_ranges = diagnostics
+
+    // Assemble a list of diagnostics by line number
+    let mut diagnostic_lines = HashMap::new();
+    for diagnostic in diagnostics {
+        let face = match diagnostic.severity {
+            Some(DiagnosticSeverity::Error) => "InlayDiagnosticError",
+            _ => "InlayDiagnosticWarning",
+        };
+        let line_diagnostics =
+            diagnostic_lines
+                .entry(diagnostic.range.end.line)
+                .or_insert(LineDiagnostics {
+                    range_end: diagnostic.range.end,
+                    symbols: String::new(),
+                    text: String::new(),
+                    text_face: "",
+                    text_severity: None,
+                });
+
+        if let Some(severity) = diagnostic.severity {
+            if line_diagnostics
+                .text_severity
+                // Smaller == higher severity
+                .map_or(true, |text_severity| severity < text_severity)
+            {
+                line_diagnostics.text = diagnostic
+                    .message
+                    .split('\n')
+                    .next()
+                    .unwrap_or_default()
+                    .replace("|", "\\|");
+                line_diagnostics.text_face = face;
+                line_diagnostics.text_severity = diagnostic.severity;
+            }
+        }
+
+        line_diagnostics
+            .symbols
+            .push_str(&format!("{{{}}}%opt[lsp_inlay_diagnostic_sign]", face))
```

Comment:
The `if` block ignores everything that doesn't have a severity, thereby dropping the message, but the statement below adds an indicator for that ignored message regardless.

Maybe we should apply the same severity default as above:
```rust
let severity = diagnostic.severity.unwrap_or(DiagnosticSeverity::Warning);
```